File: c#-solution/0788. Rotated Digits.cs
// 788. Rotated Digits
// https://leetcode.com/problems/rotated-digits/
// Tag: C#, Math, String, Medium


// Math approach
// This approach checks each digit of the number and determines if it contains valid rotated digits.
public class Solution {
    public int RotatedDigits(int n) {
        int count = 0;
        for (int i = 1; i <= n; i++) {
            if (IsValid(i)) {
                count++;
            }
        }
        return count;
    }

    private bool IsValid(int num) {
        bool hasValidDigit = false;
        while (num > 0) {
            int digit = num % 10;
            if (digit == 3 || digit == 4 || digit == 7) {
                return false; // Contains invalid digit
            }
            if (digit == 2 || digit == 5 || digit == 6 || digit == 9) {
                hasValidDigit = true; // Contains a valid rotated digit
            }
            num /= 10;
        }
        return hasValidDigit; // Must contain at least one valid rotated digit
    }
}
// TC: O(n * d) where n is the number of integers from 1 to n and d is the number of digits in each integer
// SC: O(1) since we are using a constant amount of space for the variables


// string approach
// This approach converts each number to a string and checks each character.
public class Solution {
    public int RotatedDigits(int n) {
        var invalid = new HashSet<char>{ '3', '4', '7'};
        var validDiff = new HashSet<char>{ '2', '5', '6', '9'};
        int result = 0;

        for(int i=1; i<= n; i++){
            var str = i.ToString();
            var hasDiff = false;
            var isValid = true;

            foreach(var c in str){
                if(invalid.Contains(c)){
                    isValid = false;
                    break;
                }
                if(validDiff.Contains(c)){
                    hasDiff = true;
                }
            }
            if(isValid && hasDiff) result ++;
        }

        return result;
    }
}

File: medium-problem/0788_Rotated Digits.js
// 788. Rotated Digits
// https://leetcode.com/problems/rotated-digits/
// Tag: Math, String, Brainteaser, Medium


// An integer x is a good if after rotating each digit individually by 180 degrees, we get a valid number that is different from x. Each digit must be rotated - we cannot choose to leave it alone.

// A number is valid if each digit remains a digit after rotation. For example:

// 0, 1, and 8 rotate to themselves,
// 2 and 5 rotate to each other (in this case they are rotated in a different direction, in other words, 2 or 5 gets mirrored),
// 6 and 9 rotate to each other, and
// the rest of the numbers do not rotate to any other number and become invalid.
// Given an integer n, return the number of good integers in the range [1, n].

/**
 * @param {number} n
 * @return {number}
 */

// [brute force] -AC 63ms
var rotatedDigits = function(n) {
    let conter = 0;
    for(let i=1; i<=n; i++){
        if(isvalidate(i)) conter ++
    }
    return conter
};

function isvalidate(i){
    let validate = false
    while(i>0){
      let a = i%10;
      if(a===3 || a===4|| a===7) return false
      if(a=== 2||a===5||a===6||a===9) validate = true
      i = Math.floor(i/10)
    }
    return validate
}
// TC: n log n
// MC: 1

var rotatedDigits = function(n) {
    const diff = new Set(['2', '5', '6', '9']);
    const invalid = new Set(['3', '4', '7']);

    let result = 0;
    for(let i=0; i<=n; i++){
        let str = i + "";
        let arr = str.split('');
        let hasdiff = false;
        let isValid = true;
        for(let c of arr){
            if(diff.has(c)){
                hasdiff = true;
            } else if (invalid.has(c)){
                isValid = false;
                break;
            }
        }
        if(hasdiff && isValid){
            result ++;
        }
    }
    return result;

};

console.log(rotatedDigits(857)) //247